esti/ugc_test.go

```diff
@@ -214,16 +220,177 @@ func listUnderlyingStorage(t *testing.T, ctx context.Context, s3Client *s3.S3, b
+
+	// pre run
+	getPhysicalResp, err := client.GetPhysicalAddressWithResponse(ctx, repo, mainBranch, &api.GetPhysicalAddressParams{Path: "foo/bar"})
+	if err != nil {
+		t.Fatalf("Failed to get physical address %s", err)
+	}
+	if getPhysicalResp.JSON200 == nil {
+		t.Fatalf("Failed to get physical address information: status code %d", getPhysicalResp.StatusCode())
+	}
+
+	// upload files while ugc is running
+	ticker := time.NewTicker(time.Second * 1)
+	done := make(chan bool)
+	go func() {
+		for {
+			select {
+			case <-done:
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				uploadAndDeleteSafeTestData(t, ctx, repo)
+			}
+		}
+	}()
+
+	// run ugc
+	submitConfig := &sparkSubmitConfig{
+		sparkVersion:    sparkImageTag,
+		localJar:        metaclientJarPath,
+		entryPoint:      "io.treeverse.gc.UncommittedGarbageCollector",
+		extraSubmitArgs: []string{"--conf", "spark.hadoop.lakefs.debug.gc.uncommitted_min_age_seconds=1"},
+		programArgs:     []string{uncommittedSafeGCRepoName, "us-east-1"},
+		logSource:       "ugc",
+	}
+	testutil.MustDo(t, "run uncommitted GC", runSparkSubmit(submitConfig))
+	done <- true
+
+	// post run
+	const expectedSizeBytes = 38
+	_, err = client.LinkPhysicalAddressWithResponse(ctx, "repo1", "main", &api.LinkPhysicalAddressParams{
+		Path: "foo/bar",
+	}, api.LinkPhysicalAddressJSONRequestBody{
+		Checksum:  "afb0689fe58b82c5f762991453edbbec",
+		SizeBytes: expectedSizeBytes,
+		Staging: api.StagingLocation{
+			PhysicalAddress: getPhysicalResp.JSON200.PhysicalAddress,
+			Token:           getPhysicalResp.JSON200.Token,
+		},
+	})
+	if err != nil {
+		t.Fatalf("Failed to link physical address %s", err)
+	}
+
+	s3Client := newDefaultS3Client()
+	validateSafeUncommittedGC(t, ctx, s3Client, repo)
+}
+
+func uploadAndDeleteSafeTestData(t *testing.T, ctx context.Context, repository string) {
+	name := xid.New().String()
+	_, _ = uploadFileRandomData(ctx, t, repository, mainBranch, name, false)
+
+	_, err := client.DeleteObjectWithResponse(ctx, repository, mainBranch, &api.DeleteObjectParams{Path: name})
+	if err != nil {
+		t.Fatalf("Failed to delete object %s", err)
+	}
+}
+
+func validateSafeUncommittedGC(t *testing.T, ctx context.Context, s3Client *s3.S3, repository string) {
+	objects, qp := listRepositoryUnderlyingStorage(t, ctx, repository)
+	runID := getLastUGCRunID(t, ctx, s3Client, qp.StorageNamespace, qp.Prefix)
+
+	reportPath := fmt.Sprintf("%s_lakefs/retention/gc/uncommitted/%s/summary.json", qp.Prefix, runID)
+	startListTime, err := getReportCutoffTime(s3Client, qp.StorageNamespace, reportPath)
+	if err != nil {
+		t.Fatalf("Failed to get start list time from ugc report %s", err)
+	}
+
+	// verify that all objects are deleted, if needed
+	for _, obj := range objects {
+		if strings.Contains(obj.Key, "/_lakefs/retention/gc/uncommitted/") || strings.Contains(obj.Key, "dummy") {
+			continue
+		}
+
+		if obj.LastModified.Before(startListTime) {
```

Comment:
OK now I see what this test is all about. You upload many objects and delete them all. Some are written before the cutoff and some after the cutoff. Here you validate that all objects before the cutoff are deleted which is great.
I think the previous test checks that too and also that committed/uncommitted objects remain (i.e. not deleted).
Here's what I think is still missing:
1. Validation that deleted objects that were created after the cutoff are not deleted.
2. Validation that copied objects are not deleted (i.e. copy table works). It's a difficult one, you need to copy while deleting the source object, at the exact time that the uncommitted objects are being listed for UGC, yeah..